File: Visual Studio Product Team/Official Visual Studio 2010 Samples for C# 4.0/[VB]-Official Visual Studio 2010 Samples for C# 4.0/ADO.NET Entity Framework/VB/Model/Department.vb
' Copyright © Microsoft Corporation.  All Rights Reserved.
' This code released under the terms of the 
' Microsoft Public License (MS-PL, http://opensource.org/licenses/ms-pl.html.)



Imports Microsoft.VisualBasic
Imports System
Imports System.Collections.Generic
Imports System.Collections.ObjectModel
Imports System.Diagnostics.CodeAnalysis

Namespace EmployeeTracker.Model

    ''' <summary>
    ''' Represents a Department within the company
    ''' </summary>
    Public Class Department
        ''' <summary>
        ''' The Employees that belong to this Department
        ''' </summary>
        Private privateEmployees As ICollection(Of Employee)

        ''' <summary>
        ''' Initializes a new instance of the Department class.
        ''' </summary>
        Public Sub New()
            ' Wire up the employees collection to sync references
            ' NOTE: When running against Entity Framework with change tracking proxies this logic will not get executed
            '       because the Employees property will get over-ridden and replaced with an EntityCollection<Employee>.
            '       The EntityCollection will perform this fixup instead.
            Dim emps As New ObservableCollection(Of Employee)()
            Me.privateEmployees = emps
            AddHandler emps.CollectionChanged, Sub(sender, e)
                                                   ' Set the reference on any employees being added to this department
                                                   If e.NewItems IsNot Nothing Then
                                                       For Each item As Employee In e.NewItems
                                                           If item.Department IsNot Me Then
                                                               item.Department = Me
                                                           End If
                                                       Next item
                                                   End If

                                                   ' Clear the reference on any employees being removed that still points to this department
                                                   If e.OldItems IsNot Nothing Then
                                                       For Each item As Employee In e.OldItems
                                                           If item.Department Is Me Then
                                                               item.Department = Nothing
                                                           End If
                                                       Next item
                                                   End If
                                               End Sub
        End Sub

        ''' <summary>
        ''' Gets or sets the Id of this Department
        ''' </summary>
        Public Overridable Property DepartmentId() As Integer

        ''' <summary>
        ''' Gets or sets the Name of this Department
        ''' </summary>
        Public Overridable Property DepartmentName() As String

        ''' <summary>
        ''' Gets or sets the Code of this Department
        ''' </summary>
        Public Overridable Property DepartmentCode() As String

        ''' <summary>
        ''' Gets or sets the date this Department was last audited
        ''' </summary>
        Public Overridable Property LastAudited() As DateTime?

        ''' <summary>
        ''' Gets or sets the employees that belong to this Department
        ''' Adding or removing will fixup the department property on the affected employee
        ''' </summary>
        Public Overridable Property Employees() As ICollection(Of Employee)
            Get
                Return Me.privateEmployees
            End Get
            Set(ByVal value As ICollection(Of Employee))
                Me.privateEmployees = value
            End Set
        End Property
    End Class
End Namespace

File: Visual Studio Product Team/Official Visual Studio 2010 Samples for C# 4.0/46872-Official Visual Studio 2010 Samples for C# 4.0/ADO.NET Entity Framework/C#/Model/Department.cs
// Copyright © Microsoft Corporation.  All Rights Reserved.
// This code released under the terms of the 
// Microsoft Public License (MS-PL, http://opensource.org/licenses/ms-pl.html.)


namespace EmployeeTracker.Model
{
    using System;
    using System.Collections.Generic;
    using System.Collections.ObjectModel;
    using System.Diagnostics.CodeAnalysis;

    /// <summary>
    /// Represents a Department within the company
    /// </summary>
    public class Department
    {
        /// <summary>
        /// The Employees that belong to this Department
        /// </summary>
        private ICollection<Employee> employees;

        /// <summary>
        /// Initializes a new instance of the Department class.
        /// </summary>
        public Department()
        {
            // Wire up the employees collection to sync references
            // NOTE: When running against Entity Framework with change tracking proxies this logic will not get executed
            //       because the Employees property will get over-ridden and replaced with an EntityCollection<Employee>.
            //       The EntityCollection will perform this fixup instead.
            ObservableCollection<Employee> emps = new ObservableCollection<Employee>();
            this.employees = emps;
            emps.CollectionChanged += (sender, e) =>
            {
                // Set the reference on any employees being added to this department
                if (e.NewItems != null)
                {
                    foreach (Employee item in e.NewItems)
                    {
                        if (item.Department != this)
                        {
                            item.Department = this;
                        }
                    }
                }

                // Clear the reference on any employees being removed that still points to this department
                if (e.OldItems != null)
                {
                    foreach (Employee item in e.OldItems)
                    {
                        if (item.Department == this)
                        {
                            item.Department = null;
                        }
                    }
                }
            };
        }

        /// <summary>
        /// Gets or sets the Id of this Department
        /// </summary>
        public virtual int DepartmentId { get; set; }

        /// <summary>
        /// Gets or sets the Name of this Department
        /// </summary>
        public virtual string DepartmentName { get; set; }

        /// <summary>
        /// Gets or sets the Code of this Department
        /// </summary>
        public virtual string DepartmentCode { get; set; }

        /// <summary>
        /// Gets or sets the date this Department was last audited
        /// </summary>
        public virtual DateTime? LastAudited { get; set; }

        /// <summary>
        /// Gets or sets the employees that belong to this Department
        /// Adding or removing will fixup the department property on the affected employee
        /// </summary>
        public virtual ICollection<Employee> Employees
        {
            get { return this.employees; }
            set { this.employees = value; }
        }
    }
}
